### ISSUE TITLE
-var-file parse error shows command usage rather than real error message

### ISSUE THREAD

**Issue by grodzik** [hashicorp/terraform#18500]
### Terraform Version
```
~!$ terraform -v
Terraform v0.11.7
+ provider.aws v1.28.0
```

### Debug Output
```
~!$ TF_LOG=trace terraform plan
2018/07/19 15:12:42 [INFO] Terraform version: 0.11.7  41e50bd32a8825a84535e353c3674af8ce799161
2018/07/19 15:12:42 [INFO] Go runtime version: go1.10.1
2018/07/19 15:12:42 [INFO] CLI args: []string{"/home/grodzik/.local/bin/terraform", "plan"}
2018/07/19 15:12:42 [DEBUG] Attempting to open CLI config file: /home/grodzik/.terraformrc
2018/07/19 15:12:42 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2018/07/19 15:12:42 [INFO] CLI command args: []string{"plan"}
2018/07/19 15:12:42 [DEBUG] plugin: waiting for all plugin processes to complete...
Usage: terraform plan [options] [DIR-OR-PLAN]

  Generates an execution plan for Terraform.

  This execution plan can be reviewed prior to running apply to get a
  sense for what Terraform will do. Optionally, the plan can be saved to
  a Terraform plan file, and apply can take this plan file to execute
  this plan exactly.

  If a saved plan is passed as an argument, this command will output
  the saved plan contents. It will not modify the given plan.

Options:
[...] TRUNCATED FOR THE GITHUB ISSUE
zsh: exit 1     TF_LOG=trace terraform plan
```

### Expected Behavior
Terraform should show some error stating what's wrong

### Actual Behavior
Terraform shows help for used subcommand (plan, apply, validate), exits with status code 1

### Additional Context
I've added `terraform.tf` to my project, with JSON formated variable values. When running `terraform plan/apply/validate` I've only seen help message, not an actual error. It took me 15 minutes of running terraform with multiple subcommands (init worked ok actually) to finally get once the actual error:
```
~!$ terraform plan
invalid value "terraform.tfvars" for flag -var-file-default: Error parsing terraform.tfvars: At 1:1: expected: IDENT | STRING | ASSIGN | LBRACE got: LBRACK
Usage: terraform plan [options] [DIR-OR-PLAN]

  Generates an execution plan for Terraform.
[...] TRUCATED
```
Every other run there was only help message, no error message. I haven't made any changes between first time running command after adding new file and the time I got actual error message.
I knew that there was some kind of error from my side, but running commands several times to get to know the actual error is a bit wrong.

**Comment by grodzik** [comment]
Ah, now it showed that error strangly after the help message:
```
~!$ terraform plan
Usage: terraform plan [options] [DIR-OR-PLAN]

  Generates an execution plan for Terraform.

[...] TRUNCATED

  -var-file=foo       Set variables in the Terraform configuration from
                      a file. If "terraform.tfvars" or any ".auto.tfvars"
                      files are present, they will be automatically loaded.
invalid value "terraform.tfvars" for flag -var-file-default: Error parsing terraform.tfvars: At 1:1: expected: IDENT | STRING | ASSIGN | LBRACE got: LBRACKzsh: exit 1     terraform plan
```

Seems like there is some kind of race between showing an error and help message. Also, there seems to be no new line charater after the error message, so it might be that help message (that shouldn't be displayed in this case at all) clears the line before showing itself and thus removes error message.

**Comment by mildwonkey** [comment]
HI @grodzik ,
I'm sorry you've run into this issue, thank you for submitting this. Would it be accurate to summarize this as "terraform plan does not show a useful error when it cannot parse the configuration files"?

**Comment by grodzik** [comment]
More or less that's the case, with one difference, this is not just `plan`, it's also `apply` and `validate`

**Comment by apparentlymart** [comment]
Ahh, I know why this is behaving in this strange way: we currently process `-var-file` _during_ command line option parsing, and so the inability to read a file surfaces the same way as any other command line usage error, causing the usage prompt.

Agreed that this ought to be a different sort of error. We've got some changes in this area in our dev branch for the next major release, so we may have improved on this there already, but we'll need to wait until development progresses a little further before we can test and verify since the dev branch is not currently in a stable state. We'll come back here and check on this as we get closer to release and see if it's either already fixed or if there's a relatively-easy fix we can include without adding a lot of additional scope/risk to the release.

**Comment by josephholsten** [comment]
@apparentlymart would a patch for 0.11 be useful? I don't see `terraform/command.rawFlags` getting pushed back. Or would it be most useful to write a test case?

**Comment by apparentlymart** [comment]
Given the scope of the 0.12 development branch we're keeping changes to master for 0.11 to a minimum, particularly in systems touched by that branch, because otherwise the already-complex merge will be even more so.

Unfortunately [the CLI handling has some significant changes](https://github.com/hashicorp/terraform/commit/5861dbf3fc49b19587a31816eb06f511ab861bb4#diff-90b1eddd7a48b85e97074446d8cbb80aR358) and so I don't think there'll be a way to fix this in `master` without creating a messy conflict.

The change linked there is the one I think may improve this situation for 0.12, since in the new codepath the CLI handling just gathers up the raw argument value strings and then they are _processed_ in a later step, which is in a better position to return proper error messages since it's no longer running inside a call from the `flag` package.

**Comment by apparentlymart** [comment]
I went to investigate a little deeper the current state of things in the 0.12 development branch and indeed it seems like the crucial thing for addressing this issue is already done: the processing of the content of these arguments is delayed to a later step.

At this point that later step is not yet hooked up, so we can't yet test this end-to-end but as mentioned before I've labelled this such that we'll find it again during our review/test process as we get closer to the alpha release.

**Comment by apparentlymart** [comment]
Hi all!

With the various changes I mentioned now landed in master, I verified this fix with the v0.12.0-alpha2 prerelease build. To test it, I made a `terraform.tfvars` file containing a JSON-style array `[]` just as a placeholder syntax error that seemed similar to what @grodzik was trying.

On running `terraform apply`, I got a real parse error:

```
$ terraform apply

Error: Argument or block definition required

  on terraform.tfvars line 1:
   1: []

An argument or block definition is required here.
```

This error is still not perfect because it's coming from the parser which doesn't have enough context available to talk more precisely about how `tfvars` files are formatted, but it's an improvement over the total lack of error and the even-lower-level parser errors we were seeing before, so I'm going to treat this as "good enough" for the v0.12.0 release and well continue to iterate on error message quality across the board in subsequent releases.

Since this fix is already in master, I'm going to close this out now. The fix will also be in the forthcoming v0.12.0 final release. Thanks for reporting this, and thanks for your patience while we got the groundwork laid to make this work better.
